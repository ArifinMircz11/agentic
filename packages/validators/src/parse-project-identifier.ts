import { HttpError } from '@agentic/platform-core'

import type { ParsedProjectIdentifier, ParseIdentifierOptions } from './types'
import { parseDeploymentIdentifier } from './parse-deployment-identifier'
import { coerceIdentifier } from './utils'

const projectIdentifierRe = /^@([a-z0-9-]{1,256})\/([a-z0-9-]{1,256})$/

export function parseProjectIdentifier(
  identifier?: string,
  { strict = true, errorStatusCode = 400 }: ParseIdentifierOptions = {}
): ParsedProjectIdentifier {
  const inputIdentifier = identifier

  if (!strict) {
    try {
      return parseDeploymentIdentifier(identifier, {
        strict,
        errorStatusCode
      })
    } catch {
      // ignore
    }
  }

  if (!strict) {
    identifier = coerceIdentifier(identifier)
  }

  if (!identifier?.length) {
    throw new HttpError({
      statusCode: errorStatusCode,
      message: `Invalid project identifier "${inputIdentifier}"`
    })
  }

  const match = identifier.match(projectIdentifierRe)

  if (match) {
    return {
      projectIdentifier: `@${match[1]!}/${match[2]!}`,
      projectNamespace: match[1]!,
      projectName: match[2]!
    }
  }

  throw new HttpError({
    statusCode: errorStatusCode,
    message: `Invalid project identifier "${inputIdentifier}"`
  })
}
