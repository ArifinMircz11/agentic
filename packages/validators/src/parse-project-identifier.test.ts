import { describe, expect, test } from 'vitest'

import { parseProjectIdentifier } from './parse-project-identifier'
import {
  isValidNamespace,
  isValidProjectIdentifier,
  isValidProjectName
} from './validators'

function success(...args: Parameters<typeof parseProjectIdentifier>) {
  const result = parseProjectIdentifier(...args)
  expect(result).toBeTruthy()
  expect(result!.projectIdentifier).toBeTruthy()
  expect(result!.projectNamespace).toBeTruthy()
  expect(result!.projectName).toBeTruthy()
  expect(isValidProjectIdentifier(result!.projectIdentifier)).toBe(true)
  expect(isValidProjectName(result!.projectName)).toBe(true)
  expect(isValidNamespace(result!.projectNamespace)).toBe(true)
  expect(result).toMatchSnapshot()
}

function error(...args: Parameters<typeof parseProjectIdentifier>) {
  expect(() => parseProjectIdentifier(...args)).throws()
}

describe('parseProjectIdentifier', () => {
  test('strict mode valid', () => {
    success('@username/foo')
    success('@username/foo-bar')
    success('@username/foobar123-yo')
    success('@u/foo-bar')
    success('@a/foo-bar')
    success('@foo/foobar123-yo')
    success('@foo/bar')
  })

  test('strict mode invalid', () => {
    error()
    error('')
    error('foo')
    error('foo/bar')
    error('foo/bar/baz')
    error('foo/bar/baz/qux')
    error('@foo/bar/baz/qux')
    error('@foo/bar/baz/qux/quux')
    error('@foo/bar/baz/qux/quux/corge')
    error('@foo/bar/baz/qux/quux/corge/grault')
    error('@foo/bar/baz/qux/quux/corge/grault/garply')
    error('@foo/bar/baz/')
    error('@foo')
    error('@foo/bar/')

    error('@foo-bar')
    error('@%/foo-bar')
    error('@user/foo^bar')
    error('@user@foo^bar')
    error('@username/Foo-Bar')
    error('username/foo-bar/foo')
    error('@Username/foo-bar/foo')
    error('@username/Foo-bar/foo')
    error('@username/foo-bar/')
    error('username/foo-bar')

    error('@foo_bar')
    error('@Username/foo-bar')
    error('@username/Foo-bar')
    error('@username/foo_bar')
    error('@username_/foo-bar')
    error('@username/fooBar123-yo/')

    error('//@username/foo-bar@01234567/foo')
    error('https://@username/foo-bar')
    error('https://example.com/@username/foo-bar')
  })

  test('non-strict mode valid', () => {
    success('https://gateway.agentic.so/@username/foo-bar', {
      strict: false
    })
    success('/@username/foo-bar', { strict: false })
    success('@username/foo-bar', { strict: false })
    success('https://gateway.agentic.so/@username/foo-bar', {
      strict: false
    })

    success('/@username/foo-bar', { strict: false })
    success('/@username/foo-bar', { strict: false })
    success('/@username/foo-bar', { strict: false })

    success('https://gateway.agentic.so/@username/foo-bar', {
      strict: false
    })
    success('https://gateway.agentic.so/@username/foo-bar@01234567/', {
      strict: false
    })
    success('https://gateway.agentic.so/@username/foo-bar@latest', {
      strict: false
    })

    success('https://gateway.agentic.so/@username/foo-bar@01234567/foo', {
      strict: false
    })
    success('/@username/foo-bar@01234567/foo', { strict: false })
    success('@username/foo-bar@01234567/foo/', { strict: false })
    success('https://gateway.agentic.so/@username/foo-bar@01234567/foo/', {
      strict: false
    })

    success('/@username/foo-bar@01234567/foo/', { strict: false })
    success('/@username/foo-bar@latest/foo/', { strict: false })
    success('/@username/foo-bar@dev/foo/', { strict: false })
    success('/@username/foo-bar@1.0.0/foo/', { strict: false })
    success('/@username/foo-bar/foo/', { strict: false })

    success('https://gateway.agentic.so/@username/foo-bar@01234567/foo/', {
      strict: false
    })
    success('https://gateway.agentic.so/@username/foo-bar/foo/', {
      strict: false
    })
    success('https://gateway.agentic.so/@username/foo-bar@latest/foo/', {
      strict: false
    })
  })

  test('non-strict mode invalid', () => {
    error(undefined, { strict: false })
    error('', { strict: false })
    error('https://gateway.agentic.so', { strict: false })
    error('//gateway.agentic.so', { strict: false })
    error('https://gateway.agentic.so/@username', { strict: false })
    error('https://gateway.agentic.so/call/@username/foo-bar@latest/tool', {
      strict: false
    })
    error('/call/@username/foo-bar/foo', { strict: false })
    error('//@username/foo-bar/foo', { strict: false })
    error('@username/foo-bar/foo//', { strict: false })

    error('@username/foo-bar/fooðŸ˜€', { strict: false })
    error('@username/Foo-Bar@dev/foo/', { strict: false })
    error('@username/Foo-Bar', { strict: false })
    error('@username/fooðŸ˜€', { strict: false })
  })
})
