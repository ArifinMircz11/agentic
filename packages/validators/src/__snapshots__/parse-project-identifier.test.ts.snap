// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseProjectIdentifier > non-strict mode valid 1`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 2`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 3`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 4`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 5`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 6`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 7`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 8`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 9`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 10`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 11`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
  "toolName": "foo",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 12`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
  "toolName": "foo",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 13`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
  "toolName": "foo",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 14`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
  "toolName": "foo",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 15`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
  "toolName": "foo",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 16`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
  "toolName": "foo",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 17`] = `
{
  "deploymentIdentifier": "@username/foo-bar@dev",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
  "toolName": "foo",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 18`] = `
{
  "deploymentIdentifier": "@username/foo-bar@1.0.0",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
  "toolName": "foo",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 19`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
  "toolName": "foo",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 20`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
  "toolName": "foo",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 21`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
  "toolName": "foo",
}
`;

exports[`parseProjectIdentifier > non-strict mode valid 22`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
  "toolName": "foo",
}
`;

exports[`parseProjectIdentifier > strict mode valid 1`] = `
{
  "projectIdentifier": "@username/foo",
  "projectNamespace": "username",
  "projectSlug": "foo",
}
`;

exports[`parseProjectIdentifier > strict mode valid 2`] = `
{
  "projectIdentifier": "@username/foo-bar",
  "projectNamespace": "username",
  "projectSlug": "foo-bar",
}
`;

exports[`parseProjectIdentifier > strict mode valid 3`] = `
{
  "projectIdentifier": "@username/foobar123-yo",
  "projectNamespace": "username",
  "projectSlug": "foobar123-yo",
}
`;

exports[`parseProjectIdentifier > strict mode valid 4`] = `
{
  "projectIdentifier": "@u/foo-bar",
  "projectNamespace": "u",
  "projectSlug": "foo-bar",
}
`;

exports[`parseProjectIdentifier > strict mode valid 5`] = `
{
  "projectIdentifier": "@a/foo-bar",
  "projectNamespace": "a",
  "projectSlug": "foo-bar",
}
`;

exports[`parseProjectIdentifier > strict mode valid 6`] = `
{
  "projectIdentifier": "@foo/foobar123-yo",
  "projectNamespace": "foo",
  "projectSlug": "foobar123-yo",
}
`;

exports[`parseProjectIdentifier > strict mode valid 7`] = `
{
  "projectIdentifier": "@foo/bar",
  "projectNamespace": "foo",
  "projectSlug": "bar",
}
`;
